// apeman-react-jumbotron/shim/node/ap_jumbotron_style.js
/**
 * Style for ApJumbotron.
 * @class ApJumbotronStyle
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _apemanReactStyle = require('apeman-react-style');

var _acolor = require('acolor');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** @lends ApJumbotronStyle */
var ApJumbotronStyle = function (_Component) {
  _inherits(ApJumbotronStyle, _Component);

  function ApJumbotronStyle() {
    _classCallCheck(this, ApJumbotronStyle);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(ApJumbotronStyle).apply(this, arguments));
  }

  _createClass(ApJumbotronStyle, [{
    key: 'render',
    value: function render() {
      var s = this;
      var props = s.props;

      var _ApJumbotronStyle$sty = ApJumbotronStyle.styleData(props);

      var all = _ApJumbotronStyle$sty.all;
      var small = _ApJumbotronStyle$sty.small;
      var medium = _ApJumbotronStyle$sty.medium;
      var large = _ApJumbotronStyle$sty.large;


      return _react2.default.createElement(
        _apemanReactStyle.ApStyle,
        { data: Object.assign(all, props.style),
          smallMediaData: small,
          mediumMediaData: medium,
          largeMediaData: large
        },
        props.children
      );
    }
  }]);

  return ApJumbotronStyle;
}(_react.Component);

Object.assign(ApJumbotronStyle, {
  propTypes: {
    style: _react.PropTypes.object,
    highlightColor: _react.PropTypes.string,
    backgroundColor: _react.PropTypes.string,
    titleSize: _react.PropTypes.number
  },
  defaultProps: {
    style: {},
    titleSize: 54,
    highlightColor: _apemanReactStyle.ApStyle.DEFAULT_HIGHLIGHT_COLOR,
    backgroundColor: _apemanReactStyle.ApStyle.DEFAULT_BACKGROUND_COLOR
  },
  styleData: function styleData(config) {
    var titleSize = config.titleSize;
    var highlightColor = config.highlightColor;
    var backgroundColor = config.backgroundColor;


    return {
      all: {
        '.ap-jumbotron': {
          display: 'block',
          boxSizing: 'border-box',
          width: '100%',
          padding: '32px',
          margin: '0 0 32px',
          position: 'relative',
          color: backgroundColor
        },
        '.ap-jumbotron-inner': {
          position: 'relative'
        },
        '.ap-jumbotron .ap-image': {
          backgroundColor: highlightColor
        },
        '.ap-jumbotron h1, .ap-jumbotron-title': {
          fontSize: titleSize + 'px',
          lineHeight: titleSize + 'px',
          margin: '8px 0',
          fontWeight: '100',
          wordWrap: 'break-word'
        },
        '.ap-jumbotron p, .ap-jumbotron-text': {
          margin: '4px 0'
        },
        '.ap-jumbotron-img-container': {
          position: 'absolute',
          left: 0,
          top: 0,
          right: 0,
          bottom: 0
        },
        '.ap-jumbotron-img': {
          width: '100% !important',
          height: '100% !important'
        },
        '.ap-jumbotron-img .ap-spinner': {
          display: 'none'
        },
        '.ap-jumbotron-img .ap-image': {
          opacity: 0.1,
          transition: 'opacity 100ms'
        },
        '.ap-jumbotron-img-ready .ap-image': {
          opacity: 1
        },
        '.ap-jumbotron-filter': {
          position: 'absolute',
          left: 0,
          top: 0,
          right: 0,
          bottom: 0,
          pointerEvents: 'none',
          backgroundColor: (0, _acolor.alpha)(highlightColor, 0.78)
        }
      },
      small: {},
      medium: {},
      large: {}
    };
  }
});

exports.default = ApJumbotronStyle;
