// apeman-react-jumbotron/shim/node/ap_jumbotron_title.js
/**
 * Title jumbotron
 * @class ApJumbotronTitle
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _asobj = require('asobj');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** @lends ApJumbotronTitle */
var ApJumbotronTitle = function (_Component) {
  _inherits(ApJumbotronTitle, _Component);

  function ApJumbotronTitle() {
    _classCallCheck(this, ApJumbotronTitle);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(ApJumbotronTitle).apply(this, arguments));
  }

  _createClass(ApJumbotronTitle, [{
    key: 'render',
    value: function render() {
      var s = this;
      var props = s.props;


      return _react2.default.createElement(
        'h1',
        { className: (0, _classnames2.default)('ap-jumbotron-title', props.className),
          style: Object.assign({}, props.style) },
        props.children
      );
    }
  }, {
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps, nextState) {
      var s = this;
      var props = s.props;
      var state = s.state;

      return !(0, _asobj.shallowEqual)(props, nextProps) || !(0, _asobj.shallowEqual)(state, nextState);
    }
  }]);

  return ApJumbotronTitle;
}(_react.Component);

Object.assign(ApJumbotronTitle, {

  // --------------------
  // Specs
  // --------------------

  propTypes: {},

  defaultProps: {}
});

exports.default = ApJumbotronTitle;
