// apeman-react-jumbotron/shim/node/ap_jumbotron.js
/**
 * apeman react components for jumbotron
 * @class ApJumbotron
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _apemanReactImage = require('apeman-react-image');

var _apemanReactMixinLayout = require('apeman-react-mixin-layout');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends ApJumbotron */
var ApJumbotron = _react2.default.createClass({
  displayName: 'ApJumbotron',


  // --------------------
  // Specs
  // --------------------

  propTypes: {
    /** Source url of image */
    imgSrc: _react.PropTypes.string,
    /** Alt text for image */
    imgAlt: _react.PropTypes.string,
    /** Image overlay color */
    imgFilter: _react.PropTypes.string
  },

  mixins: [_apemanReactMixinLayout.ApLayoutMixin],

  statics: {},

  getInitialState: function getInitialState() {
    return {
      imgReady: false
    };
  },
  getDefaultProps: function getDefaultProps() {
    return {
      imgSrc: null,
      imgAlt: '',
      imgFilter: null
    };
  },
  render: function render() {
    var s = this;
    var state = s.state;
    var props = s.props;
    var layouts = s.layouts;


    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)('ap-jumbotron', props.className),
        style: Object.assign({}, props.style) },
      _react2.default.createElement(
        'div',
        { className: 'ap-jumbotron-img-container' },
        _react2.default.createElement(_apemanReactImage.ApImage, { src: props.imgSrc,
          alt: props.imgAlt,
          width: layouts.image.width,
          height: layouts.image.height,
          onLoad: s.handleImgLoad,
          onError: s.handleImgError,
          scale: 'fill',
          className: (0, _classnames2.default)('ap-jumbotron-img', {
            'ap-jumbotron-img-ready': state.imgReady
          }) }),
        _react2.default.createElement('div', { className: 'ap-jumbotron-filter',
          style: {
            backgroundColor: props.imgFilter
          } })
      ),
      _react2.default.createElement(
        'div',
        { className: 'ap-jumbotron-inner' },
        props.children
      )
    );
  },


  // --------------------
  // For ApLayoutMixin
  // --------------------

  getInitialLayouts: function getInitialLayouts() {
    return {
      image: {}
    };
  },
  calcLayouts: function calcLayouts() {
    var s = this;
    var node = _reactDom2.default.findDOMNode(s);
    var rect = node.getBoundingClientRect();
    return {
      image: {
        width: rect.width,
        height: rect.height
      }
    };
  },


  // --------------------
  // Custom
  // --------------------

  handleImgLoad: function handleImgLoad() {
    var s = this;
    s.setState({
      imgReady: true
    });
  },
  handleImgError: function handleImgError() {
    var s = this;
    s.setState({
      imgReady: false
    });
  }
});

exports.default = ApJumbotron;
